### ISSUE TITLE
Updating PlugExtraData does not work consistently

### ISSUE THREAD

**Issue by wvuyk** [HomeSeer/Plugin-SDK#19]
When I use the following code:

> Dim DT As New Devices.Identification.TypeInfo With {
>                 .ApiType = Devices.Identification.EApiType.Device,
>                .Type = Devices.Identification.EDeviceType.Generic,
>                .SubType = PluginDeviceSubType.Sensor, 
>               .SubTypeDescription = "Sensor"
>               }
>          HSDevice.Address = ShortUniqueid(Sensorrow.uniqueid)
>            HSDevice.TypeInfo = DT
>           HSDevice.PlugExtraData = PED
>           hs.UpdateDeviceByRef(ref, HSDevice.Changes)`

No error occurs, but no update is done on the plugextradata information. The typInfo is set correctly.

If I use 

>hs.UpdatePropertyByRef(ref, Devices.EProperty.PlugExtraData, PED)

The plugextradata is set correctly. I would expect both would work?

**Comment by spudwebb** [comment]
From where does your PED object come from?
If you get it from the device then modify its content then reassign it using HSDevice.PlugExtraData = PED, the changes will just be ignored because the reference has not changed so we probably need a deeper comparison in AbstractHsDevice.cs line 404:

```
                if (value == _plugExtraData) {
                    Changes.Remove(EProperty.PlugExtraData);
                    return;
                }
```

A workaround for now is to create a new PED object instance like you do for TypeInfo for example

**Comment by wvuyk** [comment]
Spud,

It is a new object: as this is done during the creation of device/feature devices. I am using this statement:

`Dim PED As Devices.PlugExtraData = New Devices.PlugExtraData`

I also tried with creating a new DeviceFactory like the following code (forgot to mention this earlier), which also results in the name set correctly, but plugiinextradata not set:

```
Dim df As Devices.DeviceFactory = Devices.DeviceFactory.CreateDevice(IFACE_NAME)
Dim PED As Devices.PlugExtraData = New Devices.PlugExtraData
df.WithName(Lightrow.Name)
PED.AddNamed("light", True)
df.WithExtraData(PED)
Dim dd As Devices.NewDeviceData
dd = df.PrepareForHs
Dim ref = hs.CreateDevice(dd)
```
        `

**Comment by spudwebb** [comment]
> PED.AddNamed("light", True)

PED are now supposed to be string only. In C# if I try to AddNamed a boolean it does not even compile.

**Comment by wvuyk** [comment]
Spud,

I have change the true to "1" for now and I can confirm the update of pluginextradata works now for the `df.WithExtraData(PED)` and `hs.UpdateDeviceByRef`

This still means the `hs.UpdatePropertyByRef(ref, Devices.EProperty.PlugExtraData, PED)` allows different values as this works with the "True values". It would be better to make sure all paths are working the same?

Also, throwing an error when an invalid value is received might be helpfull for developers?

**Comment by jldubz** [comment]
@wvuyk 
How are you inspecting the PED after you have set it?  Looking through the code, it does not look like the issue is in the PED being set, but, rather, the PED being obtained and cast to the new format.  There are no checks on the format when it gets to HS for set; that is controlled within the PlugExtraData class itself.  If this is true, then reading the PED using GetPropertyByRef should return Nothing.  I will add a type check into AddNamed and AddUnNamed to ensure that a string is being passed in.

**Comment by wvuyk** [comment]
Jon,

if I set the PED with 

> hs.UpdatePropertyByRef(ref, Devices.EProperty.PlugExtraData, PED)

The PED gets set. Reading if back with HsFeature.PlugExtraData I get all entries returned, including the True/False value using

`SuspectAfterRegistration = CBool(HsFeature.PlugExtraData("suspectafterregistration")`

This is how I read it back. But when I use (changed using the True value with the ToString method)
```
Dim df As Devices.DeviceFactory = Devices.DeviceFactory.CreateDevice(IFACE_NAME)
Dim PED As Devices.PlugExtraData = New Devices.PlugExtraData
df.WithName(Lightrow.Name)
PED.AddNamed("light", True.ToString)
df.WithExtraData(PED)
Dim dd As Devices.NewDeviceData
dd = df.PrepareForHs
Dim ref = hs.CreateDevice(dd)
```
Then the HsFeature.PlugExtraData.NamedCount is 0, no Named entries are present. I use 4 to 7 items in the devices. All Named entries I use are string at this point, now even for the true value.
So I do not think it is an issue of wrong objects in the PlugExtraData, but the methods use different ways to create it?

**Comment by wvuyk** [comment]
What is the current status here? As 1.0.7 is out, but the forum still shows that ir does not function correct?

**Comment by jldubz** [comment]
@wvuyk 

This was addressed in commit https://github.com/HomeSeer/Plugin-SDK/commit/9bd6881e82624d6493e69254ae79d1e5fe80ec40 and should be resolved if you try it with v1.0.7
